### ISSUE TITLE
[Kernel Bare Bones] Hard Disk load disk error Qemu.

### ISSUE THREAD

**Issue by Nabir14** [cfenollosa/os-tutorial#278]
I followed everything step by step to 13 Kernel Bare Bones where I encountered a problem.

While running `os-image.bin` in Qemu it says:
**Booting From Hard Disk...**
**Boot Failed: Cannot boot from hard disk**

Running it with `-hda` instead of `-fda` my bootloader throws the error **Disk Error** with Hex `0x0c80`.

Also it gives this warning while linking kernel.o and kernel_entrypoint.o:
```sh
/usr/local/i386elfgcc/bin/i386-elf-ld: warning: cannot find entry symbol _start; defaulting to 00001000
```

Same code. How to fix?

**Comment by blingo77** [comment]
make sure that the kernel offset is `0x1000` in bootsect.asm and that you are loading 2 into the `dh` register before calling `disk_load`. 

also running  `qemu-system-x86_64 -fda os-image.bin` works pretty well for me.

The warning that you are getting that it's defaulting to 00001000 is just the kernel offset specified in bootsect.asm. Which is where the kernel will be found in memory. So this warning is nothing to worry about and will not cause any trouble when linking all your files together.

hopefully this helps

**Comment by Nabir14** [comment]
> make sure that the kernel offset is `0x1000` in bootsect.asm and that you are loading 2 into the `dh` register before calling `disk_load`.
> 
> also running `qemu-system-x86_64 -fda os-image.bin` works pretty well for me.
> 
> The warning that you are getting that it's defaulting to 00001000 is just the kernel offset specified in bootsect.asm. Which is where the kernel will be found in memory. So this warning is nothing to worry about and will not cause any trouble when linking all your files together.
> 
> hopefully this helps

It didn't work. I mean everything you said is correct and in my code they are too. The problem is it can't load the disk so if I use -fda it loads from floppy disk (since it can't load from hard disk) but I want it to load from hard disk so I use -hda and it returns error code 0x0c80. My disk.asm is alright.

Why don't you check my source code and see if I am doing anything wrong.
https://github.com/Nabir14/AmpereOS

**Comment by Caljay** [comment]
Not sure if you still need help, but the issue appears to be from qemu/emulation. Technically the bootloader + kernel are only 2 sectors, where the first one is loaded by the BIOS and when you are having the BIOS load more than 1 sector, there is not actually a second sector(3rd on the disk) to read. On real hardware this should work, but due to qemu it just, doesn't. This can be fixed by adding `-fda` or `-cdrom` as a flag for qemu.

Another way to fix it that I would not generally recommend is just by adding more `0x00` bytes onto the end of the bootloader and kernel.

The entry symbol problem can be fixed by adding `global _start` to the kernel_entry.asm file and define the `_start` label.

**Comment by Nabir14** [comment]
> Not sure if you still need help, but the issue appears to be from qemu/emulation. Technically the bootloader + kernel are only 2 sectors, where the first one is loaded by the BIOS and when you are having the BIOS load more than 1 sector, there is not actually a second sector(3rd on the disk) to read. On real hardware this should work, but due to qemu it just, doesn't. This can be fixed by adding `-fda` or `-cdrom` as a flag for qemu.
> 
> Another way to fix it that I would not generally recommend is just by adding more `0x00` bytes onto the end of the bootloader and kernel.
> 
> The entry symbol problem can be fixed by adding `global _start` to the kernel_entry.asm file and define the `_start` label.

Yes I still need help. Well thanks I might try this and see if it works.

**Comment by Nabir14** [comment]
I'm closing this as temporarily solved.

**Comment by Nabir14** [comment]
I have rebuild the cross-compiler and it seems to be fixed entirely. I think that this was because of the way my cross-compiler was setup and an typo issue. It's been fixed. 

Thanks, this problem is permanently fixed.
